libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/OnlineRouteAlternativesSwitch.kt

```diff
@@ -0,0 +1,328 @@
+    routeRequestMechanism: RouteRequestMechanism,
+    minimumRetryInterval: Long,
+    navigationRouteSerializationDispatcher: CoroutineDispatcher,
+    avoidManeuverSeconds: Int
+): Flow<List<NavigationRoute>> =
+    merge(
+        externalRoutesSet.map {
+            emptyList()
+        },
+        routesUpdatedEvent
+    ).mapLatest {
+        if (it.isEmpty()) {
+            null
+        } else {
+            val primaryRoute = it.first()
+            if (primaryRoute.origin == Onboard) {
+                logI(LOG_CATEGORY) { "Current route is offline, requesting online routes" }
+                requestOnlineRouteWithRetryOrNull(
+                    matchingResults,
+                    primaryRoute,
+                    routeProgressUpdates,
+                    routeRequestMechanism,
+                    navigationRouteSerializationDispatcher,
+                    minimumRetryInterval,
+                    avoidManeuverSeconds,
+                )
+            } else {
+                null
+            }
+        }
+    }.filterNotNull()
+
+private suspend fun requestOnlineRouteWithRetryOrNull(
+    matchingResults: Flow<LocationMatcherResult>,
+    primaryRoute: NavigationRoute,
+    routeProgress: Flow<RouteProgress>,
+    routeRequestMechanism: RouteRequestMechanism,
+    navigationRouteSerializationDispatcher: CoroutineDispatcher,
+    minimumRetryInterval: Long,
+    avoidManeuverSeconds: Int
+): List<NavigationRoute>? = coroutineScope {
+    while (true) {
+        val retryLimiter = launch { delay(minimumRetryInterval) }
+        try {
+            val latestMatchingResult = matchingResults.first()
+            val latestRouteProgress = routeProgress.first()
+            logI(LOG_CATEGORY) { "Requesting routes from $latestMatchingResult" }
+            val newRouteOptions = updateRouteOptionsOrNull(
+                primaryRoute,
+                latestRouteProgress,
+                latestMatchingResult
+            )?.applyAvoidManeuvers(
+                avoidManeuverSeconds,
+                latestMatchingResult.enhancedLocation.speed
+            )
+            if (newRouteOptions == null) {
+                logE(LOG_CATEGORY) {
+                    "Error calculating route options for online route, will retry later"
+                }
+                retryLimiter.join()
+                continue
+            }
+
+            val routeRequestResult = try {
+                routeRequestMechanism(newRouteOptions)
+            } catch (ce: CancellationException) {
+                throw ce
+            } catch (t: Throwable) {
+                DirectionsRequestResult.ErrorResponse.RetryableError
+            }
+            when (routeRequestResult) {
+                is DirectionsRequestResult.SuccessfulResponse -> {
+                    val navigationRoutes = withContext(navigationRouteSerializationDispatcher) {
+                        NavigationRoute.create(
+                            routeRequestResult.body,
```

Comment:
What about DataRef? We're not using it here, right? Why?

Reply:
`DataRef` definitely makes sense here. I didn't use only because `mapbox-java` doesn't expose API for writing model into a strem. Given the time pressure I decided to postpone this optimisation. Would you like to help me with mapbox-java today so that while I'm polishing this PR you release a new version?